### PR TITLE
WL-451: Clean up ansible services/roles for theming

### PR DESCRIPTION
**Pull request by saleem-latif**
Hi @mattdrayer , @feanil , @fredsmith 

Kindly review this PR it contains changes for WL-451, I have refactored out git-clone and add-user roles that are common to both edx_service and edx-themes roles.

### CODE REVIEW

File: playbooks/roles/add-user/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+# code style: https://openedx.atlassian.net/wiki/display/OpenOPS/Ansible+Code+Conventions
+# license:    https://github.com/edx/configuration/blob/master/LICENSE.TXT
+#
+#
+#
+# Tasks for role add-user
+#
+# Overview:
+#
+# This role performs the repetive tasks that most edX roles
```
Comment (mattdrayer) [inline playbooks/roles/add-user/tasks/main.yml]: Nit: "repetitive"

File: playbooks/roles/add-user/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+# require in our default configuration.
+#
+# Example play:
+#
+# Rather than being included in the play, this role
+# is included as a dependency by other roles in the meta/main.yml
+# file.  The including role should add the following
+# depency definition.
```
Comment (mattdrayer) [inline playbooks/roles/add-user/tasks/main.yml]: Nit: "dependency"

File: playbooks/roles/add-user/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    createhome=yes
+    shell=/bin/false
+    generate_ssh_key=yes
+  tags:
+    - install
+    - install:base
+
+# Assumes that the home directory has been created above.
+# In some cases(vagrant boxes) the home directory gets created
+# but does not have the corrent owner and group. In vagrant for
```
Comment (mattdrayer) [inline playbooks/roles/add-user/tasks/main.yml]: Nit:  "correct"  (or "current"?)

File: playbooks/roles/add-user/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+# example we were seeing it defaulting it to `root` for both.
```
Comment (mattdrayer) [inline playbooks/roles/add-user/tasks/main.yml]: Nit:  Not sure what the difference is between the two "it" references here

File: playbooks/roles/add-user/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+# Here we ensure that the ownership
+# of the home directory is always correct before proceeding.
+- name: ensure the ownership of home directory
```
Comment (mattdrayer) [inline playbooks/roles/add-user/tasks/main.yml]: Could rename to "ensure correct ownership of home directory"

File: playbooks/roles/add-user/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  file: >
```
Comment (mattdrayer) [inline playbooks/roles/add-user/tasks/main.yml]: @fredsmith is this syntax correct?  I recall having to change a similar version to an alternative syntax in #3034

File: playbooks/roles/add-user/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+#     dirs:
+#       - {path: /edx/var/edx-themes, owner: 'edx-themes', group: "edx-themes"}
+#       - {path: /edx/etc/edx-themes, owner: 'edx-themes', group: "edx-themes"}
+#       - ...
+#
+# Generating an ssh key so users can do a git
+# clone over ssh for public repositories without any
+# additional configuration
+- name: create application user
+  user: >
```
Comment (mattdrayer) [inline playbooks/roles/add-user/tasks/main.yml]: I believe the syntax needs to change here

File: playbooks/roles/add-user/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    path="{{ user_home }}"
+    state=directory
+    owner="{{ user_name }}"
+    group="{{ group_name }}"
+  tags:
+    - install
+    - install:base
+
+- name: create required dirs for the user
+  file: >
```
Comment (mattdrayer) [inline playbooks/roles/add-user/tasks/main.yml]: syntax

File: playbooks/roles/add-user/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    path="{{ user_home }}"
+    state=directory
+    owner="{{ user_name }}"
+    group="{{ group_name }}"
+  tags:
+    - install
+    - install:base
+
+- name: create required dirs for the user
```
Comment (mattdrayer) [inline playbooks/roles/add-user/tasks/main.yml]: what do we mean by 'required' directories?

File: playbooks/roles/edx_service/meta/main.yml [hunk 10]
```diff
@@ -18,3 +18,36 @@
...
+      - path: "{{ edx_service_home }}/venvs"
+        owner: "{{ edx_service_name }}"
+        group: "{{ common_web_group }}"
+      - path: "{{ edx_service_home }}/data"
+        owner: "{{ edx_service_name }}"
+        group: "{{ common_web_group }}"
+      - path: "{{ edx_service_home }}/staticfiles"
+        owner: "{{ edx_service_name }}"
+        group: "{{ common_web_group }}"
+
```
Comment (mattdrayer) [inline playbooks/roles/edx_service/meta/main.yml]: Why the line break here?   If it has a purpose that's okay, but if that's the case are other line breaks necessary?  If the line break has a purpose we should probably indicate what that purpose is with a code comment.  Or maybe it's just extra and can be removed 😄

File: playbooks/roles/edx_service/meta/main.yml [hunk 11]
```diff
@@ -18,3 +18,36 @@
...
+      - path: "/edx/var/{{ edx_service_name }}"
+        owner: "{{ edx_service_user }}"
+        group: "{{ common_web_group }}"
+      - path: "/edx/etc/{{ edx_service_name }}"
+        owner: "{{ edx_service_user }}"
+        group: "{{ common_web_group }}"
+
```
Comment (mattdrayer) [inline playbooks/roles/edx_service/meta/main.yml]: Another line break here...

File: playbooks/roles/git-clone/defaults/main.yml [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+#
+# edX Configuration
+#
+# github:     https://github.com/edx/configuration
+# wiki:       https://openedx.atlassian.net/wiki/display/OpenOPS
+# code style: https://openedx.atlassian.net/wiki/display/OpenOPS/Ansible+Code+Conventions
+# license:    https://github.com/edx/configuration/blob/master/LICENSE.TXT
+#
+##
+# Defaults for role git-checkout
```
Comment (mattdrayer) [inline playbooks/roles/git-clone/defaults/main.yml]: git-checkout or git-clone?

File: playbooks/roles/git-clone/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+# code style: https://openedx.atlassian.net/wiki/display/OpenOPS/Ansible+Code+Conventions
+# license:    https://github.com/edx/configuration/blob/master/LICENSE.TXT
+#
+#
+#
+# Tasks for role git-clone
+#
+# Overview:
+#
+# This role performs the repetive tasks that most edX roles
```
Comment (mattdrayer) [inline playbooks/roles/git-clone/tasks/main.yml]: repetive -> repetitive

**Comment by mattdrayer** [comment]
Looking good @saleem-latif!  This refactor has really cleaned things up and made things much more reusable for future initiatives.  I mainly had some nits and questions here and there, so as far as the validity of the actual roles+tasks I'll have to leave that to @fredsmith and @feanil.  The build is green, however, so hopefully that is a good sign :)

**Comment by saleem-latif** [comment]
@mattdrayer I have incorporated your feedback changes, kindly take a look

**Comment by mattdrayer** [comment]
👍

File: playbooks/roles/add-user/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    - install
+    - install:base
+
+- name: create dirs for the user
+  file:
+    path: "{{ item.path }}"
+    state: directory
+    owner: "{{ item.owner }}"
+    group: "{{ item.group }}"
+    mode: 0755
```
Comment (fredsmith) [inline playbooks/roles/add-user/tasks/main.yml]: We don't always want these permissions.

File: playbooks/roles/add-user/tasks/main.yml [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+  file:
+    path: "{{ user_home }}"
+    state: directory
+    owner: "{{ user_name }}"
+    group: "{{ group_name }}"
+  tags:
+    - install
+    - install:base
+
+- name: create dirs for the user
```
Comment (fredsmith) [inline playbooks/roles/add-user/tasks/main.yml:66]: I feel like this is out of scope for this role.

File: playbooks/roles/edx_service/meta/main.yml [hunk 16]
```diff
@@ -18,3 +18,34 @@
...
+      - path: "{{ edx_service_home }}/staticfiles"
+        owner: "{{ edx_service_name }}"
+        group: "{{ common_web_group }}"
+      - path: "/edx/var/{{ edx_service_name }}"
+        owner: "{{ edx_service_user }}"
+        group: "{{ common_web_group }}"
+      - path: "/edx/etc/{{ edx_service_name }}"
+        owner: "{{ edx_service_user }}"
+        group: "{{ common_web_group }}"
+      - path: "{{ COMMON_LOG_DIR }}/{{ edx_service_name }}"
```
Comment (fredsmith) [inline playbooks/roles/edx_service/meta/main.yml]: this is definately out of scope for this user creation - this directory isn't even owned by the user we're creating!

**Comment by fredsmith** [comment]
@saleem-latif - I like the idea of both roles, but I feel that the create dirs task is overreaching, and should be back in edx_service.

File: playbooks/roles/add-user/tasks/main.yml (hunk 15 above)
Comment (saleem-latif) [inline playbooks/roles/add-user/tasks/main.yml:66]: @fredsmith add-user role was created after the discussion [here](https://openedx.atlassian.net/browse/WL-451?focusedCommentId=189422&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-189422), we need this role and the dir creation tasks here because otherwise we will not be able to refactor out git-clone role from edx_service

File: playbooks/roles/edx_service/meta/main.yml (hunk 16 above)
Comment (saleem-latif) [inline playbooks/roles/edx_service/meta/main.yml]: @fredsmith I think we can move the creation of log dirs back to edx_service unless they need to be created before git-clone? I think they are not but let me know if I am wrong.

**Comment by feanil** [comment]
Part of the reason that needed to be pulled it was because the github role expects those dirs is that right @saleem-latif .  Is there any reason that all the folders can't be created with the user?

**Comment by fredsmith** [comment]
I just thought it was difficult to track down where and how a directory was created, if it's created as part of a meta dependency.  If there's the requirement that the directories be included before the github checkout runs, then I think that's OK.  perhaps we could leave directories (like the syslog log directory) that aren't directly needed by the git checkout as separate tasks?

**Comment by fredsmith** [comment]
@saleem-latif - @feanil and I discussed this, and we'd like you to take the directory creation out of the call to add-user from edx_service, and make them tasks in edx_service.  The task can stay in the add-user role, but leave it unused in this case.

Also, please make the mode for directories you're creating overridable in your dictionary.  You can make 0755 the default.

**Comment by mattdrayer** [comment]
For my own edification, why leave the task in add-user if it will not be used?

**Comment by saleem-latif** [comment]
@feanil , @fredsmith I have update code according to you feedback, and tested the changes on sandbox [here](ecommerce-conf-test-2.sandbox.edx.org). Kindly take a look now.

**Comment by mattdrayer** [comment]
@fredsmith @feanil is there anything else you would like to see in this PR?  Our current sprint is closing up tomorrow (Monday) @ 11am Eastern -- I'd love to merge this change set and close the ticket, if possible.  Let us know, thanks!

**Comment by fredsmith** [comment]
👍

**Comment by fredsmith** [comment]
👍

**Comment by saleem-latif** [comment]
okay, going ahead with the merge now

**Comment by feanil** [comment]
Looks like this was never tested on a sandbox...

TASK: [git-clone | remove read-only ssh key] **********************************
fatal: [localhost] => One or more undefined variables: 'unicode object' has no attribute 'REPO'

FATAL: all hosts have already failed -- aborting

Failed while trying to get edx-platform weekly release out.  Reverting.

**Comment by fredsmith** [comment]
looks like we should have added a:
```
when: GIT_REPOS is defined
```
 to that last task.

File: playbooks/roles/git-clone/tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+  with_items: GIT_REPOS
+  tags:
+    - install
+    - install:code
+
+- name: remove read-only ssh key
+  file:
+    dest: "{{ git_home }}/.ssh/{{ item.REPO }}"
+    state: absent
+  when: GIT_REPOS is defined
```
Comment (fredsmith) [inline playbooks/roles/git-clone/tasks/main.yml:81]: should be:

```
when: item.PROTOCOL == "ssh" and GIT_REPOS is defined
```
